go/common/crypto/scrape/s11n.go

```diff
@@ -0,0 +1,426 @@
+	return s.UnmarshalBinary(b)
+}
+
+// MarshalBinary implements the encoding.BinaryMarshaler interface.
+func (s Scalar) MarshalBinary() ([]byte, error) {
+	data, err := s.inner.MarshalBinary()
+	if err != nil {
+		return nil, fmt.Errorf("scrape/s11n: failed to serialize scalar: %w", err)
+	}
+
+	return data, nil
+}
+
+// MarshalPEM encodes a scalar into PEM form.
+func (s Scalar) MarshalPEM() ([]byte, error) {
+	b, err := s.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	return pem.Marshal(scalarPEMType, b)
+}
+
+// LoadOrGeneratePEM loads a scalar from a PEM file on disk.  Iff the
+// scalar is missing, a new one will be generated, written, and loaded.
+func (s *Scalar) LoadOrGeneratePEM(fn string) error {
+	f, err := os.Open(fn) //nolint: gosec
+	if err != nil {
+		if os.IsNotExist(err) {
+			var newScalar *Scalar
+			if newScalar, _, err = NewKeyPair(); err != nil {
+				return fmt.Errorf("scrape/s11n: failed to generate new scalar: %w", err)
+			}
+			s.inner = newScalar.inner
+
+			var buf []byte
+			if buf, err = s.MarshalPEM(); err != nil {
+				return err
+			}
+
+			return ioutil.WriteFile(fn, buf, filePerm)
+		}
+		return err
+	}
+	defer f.Close() //nolint: errcheck
+
+	buf, err := ioutil.ReadAll(f)
+	if err != nil {
+		return fmt.Errorf("scrape/s11n: failed to read PEM serialized scalar: %w", err)
+	}
+
+	if err = s.UnmarshalPEM(buf); err != nil {
+		return fmt.Errorf("scrape/s11n: failed to parse PEM serialized scalar: %w", err)
+	}
+	if err = s.isWellFormed(); err != nil {
+		return fmt.Errorf("scrape/s11n: deserialized scalar is invalid: %w", err)
+	}
+
+	return nil
+}
+
+func (s *Scalar) isWellFormed() error {
+	// Can never happen(?), but check anyway.
+	if s.inner == nil {
+		return fmt.Errorf("scrape/s11n: scalar is missing")
+	}
+
+	return nil
+}
+
+func scalarFromKyber(s kyber.Scalar) Scalar {
+	return Scalar{
+		inner: s,
+	}
+}
+
+// PubVerShare is a public verifiable share (`pvss.PubVerShare`)
+type PubVerShare struct {
+	V Point `json:"v"` // Encrypted/decrypted share
+
+	C  Scalar `json:"c"`  // Challenge
+	R  Scalar `json:"r"`  // Response
+	VG Point  `json:"vg"` // Public commitment with respect to base point G
+	VH Point  `json:"vh"` // Public commitment with respect to base point H
+}
+
+func (pvs *PubVerShare) isWellFormed() error {
+	if err := pvs.V.isWellFormed(); err != nil {
+		return fmt.Errorf("scrape/s11n: invalid PubVerShare V: %w", err)
+	}
+	if err := pvs.C.isWellFormed(); err != nil {
+		return fmt.Errorf("scrape/s11n: invalid PubVerShare C: %w", err)
+	}
+	if err := pvs.R.isWellFormed(); err != nil {
+		return fmt.Errorf("scrape/s11n: invalid PubVerShare R: %w", err)
+	}
+	if err := pvs.VG.isWellFormed(); err != nil {
+		return fmt.Errorf("scrape/s11n: invalid PubVerShare VG: %w", err)
+	}
+	if err := pvs.VH.isWellFormed(); err != nil {
+		return fmt.Errorf("scrape/s11n: invalid PubVerShare VH: %w", err)
+	}
+
+	return nil
+}
+
+func (pvs *PubVerShare) toKyber(index int) *pvss.PubVerShare {
```

Comment:
precondition `pvs.isWellFormed() == nil`

Reply:
Sure, and this is checked already...